packages/eslint-plugin/src/rules/strict-boolean-expressions.ts

```diff
@@ -275,6 +315,87 @@ export default createRule<Options, MessageId>({
       if (assertedArgument != null) {
         traverseNode(assertedArgument, true);
       }
+      if (isArrayMethodCallWithPredicate(context, services, node)) {
+        const predicate = node.arguments.at(0);
+
+        if (predicate) {
+          checkArrayMethodCallPredicate(predicate);
+        }
+      }
+    }
+
+    /**
+     * Dedicated function to check array method predicate calls. Reports predicate
+     * arguments that don't return a boolean value.
+     *
+     * Ignores the `allow*` options and requires a boolean value.
+     */
+    function checkArrayMethodCallPredicate(
+      node: TSESTree.CallExpressionArgument,
+    ): void {
+      // custom message for accidental `async` function expressions
+      if (isFunctionExpressionNode(node) && node.async) {
+        return context.report({
+          node,
+          messageId: 'predicateCannotBeAsync',
+        });
+      }
+
+      const type = checker.getApparentType(services.getTypeAtLocation(node));
```

Comment:
wondering if there's a way to leverage `checker.getResolvedSignature` or similar to deal with some edge cases around generic/overload resolution 🤔. If not, I think it's probably ok, since the arrow function form does work correctly, and is much more common in real world use cases. 

```ts
declare function f(x: number): string;
declare function f(x: string | null): boolean;


// shouldn't report; doesn't.
['one', null].filter(f);
['one', null].filter(x => f(x));


// (technically) should report; doesn't.
[35].filter(f);
// correct with arrow function form
[35].filter(x => f(x));


declare function generic<T>(x: T): T;
// (technically) shouldn't report; does.
[true, false].filter(generic);
// correct with arrow function form
[true, false].filter(x => generic(x));

// important test case with generic type predicate (important for code that predates inferred type predicates)
declare function notNullish<T>(x: T): x is NonNullable<T>;
// should not report; doesn't
['one', null].filter(notNullish);
```

***

if we can't figure out which overload will be used, I think the only change I would make is that we should probably error if _any_ of the overloads doesn't return a boolean, rather than only if _all_ don't return a boolean. Maybe that's a bit overbearing, but I think it's preferable to have always false positives here, rather than sometimes false positives _and_ sometimes false negatives

Reply:
That's an interesting problem. I looked around, but I don't think `checker.getResolvedSignature` can help here, potentially the currently not exposed [`isSignatureAssignableTo`](https://github.com/microsoft/TypeScript/blob/b3c67d32020a3015fe068fbcb1dcfb18aedcaad3/src/compiler/checker.ts/#L21273-L21275) or [`compareSignaturesRelated`](https://github.com/microsoft/TypeScript/blob/b3c67d32020a3015fe068fbcb1dcfb18aedcaad3/src/compiler/checker.ts/#L21294-L21430).

The best I can think of is to mimic that somewhat: Since we know for a fact [every array predicate method](https://github.com/typescript-eslint/typescript-eslint/blob/244435126619afb9497ace04cbf4819012e27330/packages/eslint-plugin/src/util/isArrayMethodCallWithPredicate.ts#L13-L21) takes the type of the array as its first argument (and a `number` as the second one), we can try and rule out which signature we should check.

I did manage to implement the above, so these both correctly report:

```ts
// (technically) should report; doesn't.
[35].filter(f);
// correct with arrow function form
[35].filter(x => f(x));
```

I'm not sure, however how to tackle the generic one:

```ts
declare function generic<T>(x: T): T;
// (technically) shouldn't report; does.
[true, false].filter(generic);
// correct with arrow function form
[true, false].filter(x => generic(x));
```

Wdyt? I'll add the missing test cases regardless of how this is tackled.

Reply:
> That's an interesting problem. I looked around, but I don't think checker.getResolvedSignature can help here, potentially the currently not exposed [isSignatureAssignableTo](https://github.com/microsoft/TypeScript/blob/b3c67d32020a3015fe068fbcb1dcfb18aedcaad3/src/compiler/checker.ts/#L21273-L21275) or [compareSignaturesRelated](https://github.com/microsoft/TypeScript/blob/b3c67d32020a3015fe068fbcb1dcfb18aedcaad3/src/compiler/checker.ts/#L21294-L21430).

Oh well. Thanks for looking! 

I think that I would prioritize the test cases as follows
```ts
declare function notNullish<T>(x: T): x is NonNullable<T>;

declare function f(x: number): string;
declare function f(x: string | null): boolean;

declare function generic<T>(x: T): T;

// Priority 1. The only one that is truly important.
// should not report
['one', null].filter(notNullish);

// Priority 2
// should report
[35].filter(f);

// Priority 10
// shouldn't report (but erring on the side of reporting is fine)
['one', null].filter(f);

// Priority 11
// shouldn't report (but erring on the side of reporting is fine)
[true, false].filter(generic);
```

If the heuristic implementation is simple, we can do it, but otherwise I'm not sure its worth extra code complexity. 

Does that help?

Reply:
Yes, thanks!

I've implemented this as simple as I could, and this does solve every case you mentioned but the last (which also shows as an incorrect report on `no-unnecessary-condition`), wdyt?

If you think this adds redundant complexity I'll be happy to revert this and ask all signatures to return a `boolean`.

Reply:
Notice I've pushed one additional change just now that simplifies things a bit further.

Reply:
@kirkwaiblinger note that a similar behavior can be seen on `no-unnecessary-condition` ([playground link](https://typescript-eslint.io/play/#ts=4.8.4&fileType=.tsx&code=CYUwxgNghgTiAEAzArgOzAFwJYHtXwxAGcMBGACgA8AueVZAWwCMQZ4AfO5CCASlowxkIANwAoUJFgIU6bHgLEyVWiRhZUAc37w1GzeLEB6I7oAWObsCRQsEADTwoRAmYRgcMOJl1ZNqKAxkOHgsFww3eDwEOCCYVH14AANBYSSxAG0AZgBWR3oeAF0AOkQ7QhhyQhJSXkMTGzsXDy9wDAgATycXKCQ0TFx8EEoABzgiIkHM3PzuCBKyiAqqeABeAD5FGqpeOrEgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)):

```ts
declare function test1(x: number | null): true;
declare function test1(x: string): string;

// should fail, as the correct signature is the one returning `true`
[35, null].filter(test1);

// fails correctly as a function expression
[35, null].filter(x => test1(x));
```

And also on `no-misused-promises` ([playground link](https://typescript-eslint.io/play/#ts=4.8.4&fileType=.tsx&code=CYUwxgNghgTiAEAzArgOzAFwJYHtXwxAGcMBGACgA8AueVZAWwCMQZ4AfO5CCASlqY4cEEFFQBuAFChIsBCnTY8BYmSq0SMLKgDm-eAAUYOBliIgAPPWasAfFMkB6R-CIALHN2DwADlCJEADTw-gRuCGA4MHCYrlg6qFAYyHDwZmEIeAhwyTCo2jrwAAaCwqKoRZIA2gDMAKzB9DwAugB0iFgQhDDkhCSkvA7Ovv7mRPCR0eAYEACeIeNQSGiYuPgglD5wAWvV9Y3cEG0dXaxU8AC8tir9VLyDkkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oFtLlZkiACa1i0Dr0GoMkRNHHRI4MAF8QKoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)):

```ts
declare function test1(x: number | null): boolean;
declare function test1(x: string): Promise<number>;

// should pass, as the correct signature is the one returning `boolean`
[35, null].filter(test1);

// passes correctly as a function expression
[35, null].filter(x => test1(x));
```

Since all of the examples above check call signatures of array predicates, they can reuse the heuristic I wrote to fix these bugs. If this makes sense, I'll be happy to open issues (followed by PRs) for this.